### DISCUSSION TITLE
CRA5 + TW3 app fails to build due to Invalid mapping error in Css Minimizer plugin

### DISCUSSION THREAD

**Discussion by stephent** [tailwindlabs/tailwindcss#7688] [Help] [answered]
**What version of Tailwind CSS are you using?**

3.0.23

**What build tool (or framework if it abstracts the build tool) are you using?**

"react-scripts": "5.0.0",
"autoprefixer": "^10.4.2",
"postcss": "^8.4.7"

**What version of Node.js are you using?**

14.17.5

**What browser are you using?**

Chrome 98

**What operating system are you using?**

macOS 12.1

**Reproduction URL**

https://github.com/stephent/cra-5-tw-3-build-fails

**Describe your issue**

On running `npm run build` against the reproduction project linked above, the build fails with the following error:

```
Failed to compile.

static/css/main.4bacc579.css from Css Minimizer plugin
Error: Invalid mapping: {"generated":{"line":3,"column":8651},"source":"static/css/main.4bacc579.css","original":{"line":330,"column":null},"name":null}
```

The project includes a file `./styles/custom.css` with a single class, `.my-class`. If you edit line 2 and delete the applied rule `bg-white`, the project will build without a problem.

I tried a few different rules here and only managed to reproduce the build error when a background color rule was used.

I'm unclear if this problem lies in Tailwind, CRA 5, Webpack 5, CssMinimizerPlugin or some other web pack plugin, but I figured this would be the best place to make an initial bug report.

**Comment by stephent** [comment]
@adamwathan genuine question: given that the problem only arises when using Tailwind's `@apply` directive together with a Tailwind background color utility class, and not when inlining the style, is there something else that makes you think this is not in fact an issue/bug?

e.g. this fails to build:

```
.my-class {
  @apply bg-white bg-opacity-90;
}
```

but this builds just fine:
```
.my-class {
  --tw-bg-opacity: 1;
  background-color: rgb(255 255 255 / var(--tw-bg-opacity));
  @apply bg-opacity-90;
}
```

**Comment by adamwathan** [comment]
Hey! I originally converted this because it seemed like it couldn’t really be our fault since the error isn’t thrown by Tailwind but thinking about it more maybe we aren’t generating the source map correctly or something and this error is coming from that?

Tried to reopen original issue but wouldn’t let me do it, we can track this here anyways though, I follow our discussions area pretty closely.

Can you try our insiders build, `tailwindcss@insiders`? We’ve changed a few source map things recently, maybe will see a different result.

**Reply by stephent** [reply]
Hi Adam - no change when using `tailwindcss@insiders`:

```
Failed to compile.

static/css/main.be719d6c.css from Css Minimizer plugin
Error: Invalid mapping: {"generated":{"line":3,"column":8632},"source":"static/css/main.be719d6c.css","original":{"line":330,"column":null},"name":null}
```

Let me know if there's anything else it would be useful to try.

(I'm using the `@apply` directive here in my main project to customize 3rd party components - not sure if there's any good alternative approach for that situation, short of regressing to old-school CSS.)

**Reply by johnrobertmcc** [reply]
Hi @adamwathan -> I'm having similar issues. If you want another repo to replicate it on:

Reproducible instructions:

- clone repo `https://github.com/johnrobertmcc/jrcoin`
- `npm run build`

The following error shows in terminal on build of `main`:
```
Error: Invalid mapping: {"generated":{"line":3,"column":3836},"source":"static/css/main.7cb86d24.css","original":{"line":587,"column":null},"name":null}
```

I can open a new issue if that helps; just didn't want to spam the issue log since this one is fairly new as well.

I have found that `npm run eject` and commenting out `CssMinimizerPlugin()` inside of `webpack.config.js` does allow me to build: Unsure if this is due to Tailwind. Hopeful that someone smarter can take this information and make sense of it. Branch `hotfix/eject-script` does precisely that; running build should work if that branch is pulled.

**Reply by dariusbiro** [reply]
Hi @adamwathan! Similar issues here! 

I have a slightly older project that was created with CRA (I created it maybe one year ago), but at the time the latest tailwindcss version was V2. I had to use the `postcss7-compat` version to make it work. 

Last week I upgraded to V3 and locally everything was working fine, but when I tried to deploy the app I got this error:
> static/css/main.40ea086d.css from Css Minimizer plugin
> Error: Invalid mapping: {"generated":{"line":1,"column":24261},"source":"static/css/main.40ea086d.css","original": {"line":2604,"column":null},"name":null}

The repo is private, so I can't help with that, but I just wanted to let you you know that I'd be happy to help testing the fixes.

**Reply by SleeplessByte** [reply]
@johnrobertmcc workaround here: https://github.com/tailwindlabs/tailwindcss/discussions/7688#discussioncomment-2291141

**Comment by thecrypticace** [comment]
Wow this one was a doozy. Unfortunately, I wasn't able to track down the source of the actual issue. However, I can get this to fail without Tailwind CSS involved at all.

Reproduction here: https://github.com/thecrypticace/cra-css-minimize-bug

I've removed Tailwind CSS from the project, completely emptied out index.css and replaced custom.css with the following which is what is generated by the code using `@apply` (if this looks slightly off — more on this below):

```css
.my-class {
  background-color: rgb(255 255 255 / var(--tw-bg-opacity));
  --tw-bg-opacity: 0.9
}
```

This breaks with the same error (well the line number is different but the general message is the same). Even weirder is if you flip the variable and background colors around the error disappears. It also goes away if you add a semicolon after the `0.9`. Additionally, If I disable the "CSS Declaration Sorter" cssnano plugin (which is used by the CSS minimizer in CRA) the error goes away but the underlying problem itself does not.

The core of the issue is that something is causing the internal postcss node's *end* column & offset to become corrupted (`node.source.end.offset` and `node.source.end.column`). More specifically the `offset` becomes undefined and the column becomes `NaN`.  However, this happens after Tailwind CSS generates the nodes and passes them off to PostCSS. So it's outside of our control. This means it's not a Tailwind CSS bug but we unfortunately can end up generating CSS in some cases which triggers the issue.

Aside: It appears that the `undefined` + `NaN` causes the source map processing to become order sensitive in some manner which is why disabling the declaration sorter plugin makes the error disappear.

---

Short aside about the CSS generated by `@apply`:

The following CSS:
```css
.my-class {
  @apply bg-white bg-opacity-90;
}
```

Is turned into this by Tailwind CSS when handling the `@apply` directive:
```css
.my-class {
  --tw-bg-opacity: 1;
  background-color: rgb(255 255 255 / var(--tw-bg-opacity));
  --tw-bg-opacity: 0.9
}
```

However, Tailwind CSS does internal de-duplication of declarations before outputting in some very limited cases so we eliminate the redundant variable assignment and output the following:

```css
.my-class {
  background-color: rgb(255 255 255 / var(--tw-bg-opacity));
  --tw-bg-opacity: 0.9
}
```

**Reply by stephent** [reply]
@thecrypticace many thanks for digging into this. I had also opened https://github.com/facebook/create-react-app/issues/12101 to track this, on the assumption it may not have been a Tailwind issue. I'll update that to confirm the same.

Any workarounds in how we write CSS and/or Tailwind `@apply` rules, do you think?

**Reply by thecrypticace** [reply]
Only thing I can think of is a dummy declaration after the `@apply`:
```css
.my-class {
  @apply bg-white bg-opacity-90;
  box-sizing: border-box;
}
```

When using preflight `box-sizing: border-box;` is the default for everything so that ends up being a no-op and prevents the build from failing. (you could also just use anything like `foo: bar`). Source maps are likely still bugged though.

**Reply by stephent** [reply]
One more update - it looks like some sort of regression in `postcss`. If I revert to using `"postcss": "8.4.5"` my reproduction project builds successfully. I'll look to open an issue there.

**Reply by VladTzaru** [reply]
Reverting to `postcss": "8.4.5"` did not work for me. However adding a dummy declaration like @thecrypticace suggested did! Hopefully, it's getting a fix, adding `box-sizing` everywhere is a bit strange. :)

**Reply by stephent** [reply]
@VladTzaru I think it may be due installation sequence and 8.4.7 being required by other dependencies. Try `npm ls postcss` to see what's actually being used. I found that deleting `/node_modules` and re-running `npm install` got things into a state where the project will build consistently.

**Reply by SleeplessByte** [reply]
@stephent thanks, that worked!
@VladTzaru if you're using `yarn`, you'll need to add `resolutions` for each instance of `postcss` that isn't matching (check `yarn.lock`). 

In my case it was `tailwindcss` itself that pushed `postcss` to the next patch version. Adding the following resolved it:
```json
{
  "resolutions": {
    "react-scripts/css-loader/postcss": "8.4.5",
    "react-scripts/css-minimizer-webpack-plugin/postcss": "8.4.5",
    "react-scripts/postcss": "8.4.5",
    "react-scripts/tailwindcss/postcss": "8.4.5",
    "tailwindcss/postcss": "8.4.5"
  }
}
```

If you do this, you _should_ see a single warning (if you're using tailwind 3+):

> warning Resolution field "postcss@8.4.5" is incompatible with requested version "postcss@^8.4.6"

You can tell that this now really uses the right postcss version by looking for all entries `postcss@`. As you can see, this has re-resolved all versions to `8.4.5`:

```
postcss@8.4.5, postcss@^8.3.5, postcss@^8.4.4, postcss@^8.4.5, postcss@^8.4.6:
  version "8.4.5"
  resolved "https://registry.yarnpkg.com/postcss/-/postcss-8.4.5.tgz#bae665764dfd4c6fcc24dc0fdf7e7aa00cc77f95"
  integrity sha512-jBDboWM8qpaqwkMwItqTQTiFikhs/67OYVvblFFTM7MrZjt6yMKd6r2kgXizEbTTljacm4NldIlZnhbjr84QYg==
  dependencies:
    nanoid "^3.1.30"
    picocolors "^1.0.0"
    source-map-js "^1.0.1"

postcss@^7.0.35:
  version "7.0.39"
  resolved "https://registry.yarnpkg.com/postcss/-/postcss-7.0.39.tgz#9624375d965630e2e1f2c02a935c82a59cb48309"
  integrity sha512-yioayjNbHn6z1/Bywyb2Y4s3yvDAeXGOyxqD+LnVOinq6Mdmd++SW2wUNVzavyyHxd6+DxzWGIuosg6P1Rj8uA==
  dependencies:
    picocolors "^0.2.1"
    source-map "^0.6.1"
```

**Reply by stephent** [reply]
So that's how 'resolutions' actually works! I didn't realize you had to (or even could) scope it in that way.

**Reply by SleeplessByte** [reply]
Yes, and the reason you don't do `**/postcss` is because then you would also override that 7.0.35 one.

**Comment by thecrypticace** [comment, marked as answer]
The fix for this has been released in postcss 8.4.8.

**Reply by Mobashir-Monim** [reply]
I've updated to postcss 8.4.8, but it's still not working for me
:'(

**Reply by yoloOnTheBattlefield** [reply]
same

**Reply by stephent** [reply]
You might want to try `npm ls postcss` to confirm that there's not some other dependency that is causing `postcss` to resolve to either 8.4.6 or 8.4.7.  Deleting `/node_modules` and reinstalling might also help. 

8.4.8 definitely fixed the issue for me.

**Reply by Mobashir-Monim** [reply]
I've tried doing that, but still getting the same result.

One thing that I did find is that the resolve-url-loader@4.0.0 (a dependency of react-scripts) uses postcss@7.0.39, I'm not sure if that if that is causing the error though.

I've also tried updating everything which updates postcss to 8.4.11 (leaving resolver-url-loader), again same results.
